### PR TITLE
Hybrid as default Find mode

### PR DESCRIPTION
**Pull request by daniellowell**
Set hybrid as the default, and added documentation.

### CODE REVIEW

File: doc/src/find_and_immediate.md [hunk 1]
```diff
@@ -158,3 +158,18 @@ If the user's architecture is not listed above they will need to run the Find AP
 ### Backend Limitations
 
 OpenCL support for immediate mode via the fallback is limited to fp32 datatypes. This is because this current release's fallback path goes through GEMM which on the OpenCL is serviced through MIOpenGEMM -- which itself only contains support for fp32. The HIP backend uses rocBLAS as its fallback path which contains a richer set of datatypes.
+
+
+### Find Modes
+
+MIOpen provides a set of Find modes which are used to accelerate the Find calls. The different modes are set by using the environment variable `MIOPEN_FIND_MODE`, and setting it to one of the values:
+
+- `NORMAL`: Normal Find: This is the full Find mode call, which will benchmark all the solvers and return a list.
```
Comment (atamazov) [inline doc/src/find_and_immediate.md:167]:
```suggestion
- `NORMAL` or `1`: Normal Find: This is the full Find mode call, which will benchmark all the solvers and return a list.
```

File: doc/src/find_and_immediate.md [hunk 2]
```diff
@@ -158,3 +158,18 @@ If the user's architecture is not listed above they will need to run the Find AP
...
+- `FAST`: Fast Find: Checks the [Find-Db](https://rocmsoftwareplatform.github.io/MIOpen/doc/html/finddb.html) for an entry. If there is a Find-Db hit, use that entry. If there is a miss, utilize the Immediate mode fallback. If Start-up times are expected to be faster, but worse GPU performance.
```
Comment (atamazov) [inline doc/src/find_and_immediate.md:168]:
```suggestion
- `FAST` or `2`: Fast Find: Checks the [Find-Db](https://rocmsoftwareplatform.github.io/MIOpen/doc/html/finddb.html) for an entry. If there is a Find-Db hit, use that entry. If there is a miss, utilize the Immediate mode fallback. If Start-up times are expected to be faster, but worse GPU performance.
```

File: doc/src/find_and_immediate.md [hunk 3]
```diff
@@ -158,3 +158,18 @@ If the user's architecture is not listed above they will need to run the Find AP
...
+- `HYBRID`: Hybrid Find: Checks the [Find-Db](https://rocmsoftwareplatform.github.io/MIOpen/doc/html/finddb.html) for an entry. If there is a Find-Db hit, use that entry. If there is a miss, use the existing Find machinery. Slower start-up times but none GPU performance drop at all.
```
Comment (atamazov) [inline doc/src/find_and_immediate.md:169]:
```suggestion
- `HYBRID` or `3` or variable is unset: Hybrid Find: Checks the [Find-Db](https://rocmsoftwareplatform.github.io/MIOpen/doc/html/finddb.html) for an entry. If there is a Find-Db hit, use that entry. If there is a miss, use the existing Find machinery. Slower start-up times but none GPU performance drop at all.
```

Comment (JehandadKhan) [reply doc/src/find_and_immediate.md:169]: also `none` to `no`

Comment (atamazov) [reply doc/src/find_and_immediate.md:169]: Yes. And `Slower start-up times but...` -> `Slower start-up times than Fast Find but...`

File: doc/src/find_and_immediate.md [hunk 4]
```diff
@@ -158,3 +158,18 @@ If the user's architecture is not listed above they will need to run the Find AP
 ### Backend Limitations
 
 OpenCL support for immediate mode via the fallback is limited to fp32 datatypes. This is because this current release's fallback path goes through GEMM which on the OpenCL is serviced through MIOpenGEMM -- which itself only contains support for fp32. The HIP backend uses rocBLAS as its fallback path which contains a richer set of datatypes.
+
+
+### Find Modes
+
+MIOpen provides a set of Find modes which are used to accelerate the Find calls. The different modes are set by using the environment variable `MIOPEN_FIND_MODE`, and setting it to one of the values:
```
Comment (pfultz2) [inline doc/src/find_and_immediate.md:165]: Since MIOpen's API doesnt support changing the find mode, this section should go under the debugging section.

Comment (daniellowell) [reply doc/src/find_and_immediate.md:165]: That is true. I'll link to that from here as well.

Comment (atamazov) [reply doc/src/find_and_immediate.md:165]: Please note that this is not a debugging functionality, but performance optimization, and it affects the behavior of Find API.

File: src/binary_cache.cpp [hunk 5]
```diff
@@ -159,7 +159,9 @@ void SaveBinary(const std::string& hsaco,
 
     std::string filename = (is_kernel_str ? miopen::md5(name) : name) + ".o";
     KernelConfig cfg{filename, args, hsaco};
-    MIOPEN_LOG_I2("Saving binary for: " << name << " ;args: " << args);
+    MIOPEN_LOG_I2("Saving binary for: " << (is_kernel_str ? std::string("tinygemm.cl") : name)
```
Comment (atamazov) [inline src/binary_cache.cpp:162]: It prints the whole opencl source otherwise (in case of miopengemm ;)

**Comment by atamazov** [comment]
`Full long tests` -> `Gcc Release all` failed twice in a row. Will look there when time permits.

**Comment by atamazov** [comment]
It would be nice to cherry-pick 4f498fdf64e02491a39e5a08c476eba9da18d9d1 and 564b41e73973d77598eb5af27d6d6fa402388570 to develop separately.

**Comment by daniellowell** [comment]
> Please revert all merges except those coming from the develop branch.

No, I'll merge in the PR #286 first, then this PR.

**Comment by JehandadKhan** [comment]
@daniellowell Why are all the changes to the Jenkins file necessary ?

**Comment by daniellowell** [comment]
There aren't actually a lot of changes, it is the way git is merging the file. I added an auxilery parallel stage since we are adding a bunch of specialized tests for new sub-features. This is captured in #286

File: Jenkinsfile [hunk 6]
```diff
@@ -304,6 +304,42 @@ pipeline {
                     }
                 }
 
+                stage('Hip Normal Find Mode Release') {
```
Comment (atamazov) [inline Jenkinsfile:307]: [Recommendation] I guess it is worth to mention that only convolutions are under testing in this stage.

File: Jenkinsfile [hunk 7]
```diff
@@ -304,6 +304,42 @@ pipeline {
...
+                            make -j test_conv2d
+                            MIOPEN_FIND_MODE=1 CTEST_PARALLEL_LEVEL=4 MIOPEN_DEBUG_IMPLICIT_GEMM_NON_XDLOPS_INLINE_ASM=0 MIOPEN_CONV_PRECISE_ROCBLAS_TIMING=0 bin/test_conv2d --disable-verification-cache
+                        """
+                    }
+                    steps{
+                        buildHipClangJob('/opt/rocm/llvm/bin/clang++', '', "",  image+'-hip-clang', "/usr/local", cmd)
+                    }
+                }
+
+                stage('Hip Fast Find Mode Release') {
```
Comment (atamazov) [inline Jenkinsfile:325]: Ditto

**Comment by atamazov** [comment]
@daniellowell BTW we can make the normal Find the default for DEV builds.

**Pull request merged**